src/core/processing/qgsprocessingutils.cpp

```diff
@@ -721,6 +721,14 @@ QgsFeatureIterator QgsProcessingFeatureSource::getFeatures( const QgsFeatureRequ
   return mSource->getFeatures( req );
 }
 
+QgsFeatureSource::FeatureAvailability QgsProcessingFeatureSource::hasFeatures() const
+{
+  if ( !mTransformErrorCallback && !mInvalidGeometryCallback && mInvalidGeometryCheck == QgsFeatureRequest::GeometryNoCheck )
```

Comment:
We should never be fetching geometry during a has features check, so this method should just be

    return mSource->hasFeatures();

Reply:
Ah, I see where you're going here.

The two callback checks can be skipped - they can't be used to filter features, just handle when they are encountered. 

An optimised method could be:

    FeatureAvailability sourceAvailability = mSource->hasFeatures();
    if ( sourceAvailability == NoFeaturesAvailable )
         return NoFeaturesAvailable; // never going to be features if underlying source has no features
    else if ( mInvalidGeometryCheck == QgsFeatureRequest::GeometryNoCheck )
         return sourceAvailability;
    else
         // we don't know... source has features, but these may be filitered out by invalid geometry check
         return FeaturesMaybeAvailable;

Reply:
Sounds good, thanks